// kepler.gl/dist/components/side-panel/layer-panel/color-range-selector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  padding: 12px 12px 0 12px;\n'], ['\n  padding: 12px 12px 0 12px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  padding-bottom: 12px;\n'], ['\n  padding-bottom: 12px;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  margin-bottom: 8px;\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  .color-palette__config__label {\n    flex-grow: 1;\n  }\n  .color-palette__config__select {\n    flex-grow: 1;\n  }\n  .item-selector .item-selector__dropdown {\n    ', ';\n  }\n'], ['\n  margin-bottom: 8px;\n  display: flex;\n  justify-content: space-between;\n  align-items: center;\n  .color-palette__config__label {\n    flex-grow: 1;\n  }\n  .color-palette__config__select {\n    flex-grow: 1;\n  }\n  .item-selector .item-selector__dropdown {\n    ', ';\n  }\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  padding: 0 8px;\n  :hover {\n    background-color: ', ';\n    cursor: pointer;\n  }\n'], ['\n  padding: 0 8px;\n  :hover {\n    background-color: ', ';\n    cursor: pointer;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _lodash = require('lodash.uniq');

var _lodash2 = _interopRequireDefault(_lodash);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _itemSelector = require('../../common/item-selector/item-selector');

var _itemSelector2 = _interopRequireDefault(_itemSelector);

var _styledComponents3 = require('../../common/styled-components');

var _rangeSlider = require('../../common/range-slider');

var _rangeSlider2 = _interopRequireDefault(_rangeSlider);

var _switch = require('../../common/switch');

var _switch2 = _interopRequireDefault(_switch);

var _colorPalette = require('./color-palette');

var _colorPalette2 = _interopRequireDefault(_colorPalette);

var _colorRanges = require('../../../constants/color-ranges');

var _dataUtils = require('../../../utils/data-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ALL_TYPES = (0, _lodash2.default)(_colorRanges.COLOR_RANGES.map(function (c) {
  return c.type;
}).concat(['all']));
var ALL_STEPS = (0, _lodash2.default)(_colorRanges.COLOR_RANGES.map(function (d) {
  return d.colors.length;
})).sort(_dataUtils.numberSort);

var StyledColorConfig = _styledComponents2.default.div(_templateObject);

var ColorRangeSelector = _styledComponents2.default.div(_templateObject2);
var ColorRangeSelect = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(ColorRangeSelect, _Component);

  function ColorRangeSelect() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ColorRangeSelect);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ColorRangeSelect.__proto__ || Object.getPrototypeOf(ColorRangeSelect)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      config: {
        type: {
          type: 'select',
          value: 'all',
          options: ALL_TYPES
        },
        steps: {
          type: 'select',
          value: 6,
          options: ALL_STEPS
        },
        reversed: {
          type: 'switch',
          value: false,
          options: [true, false]
        }
      }
    }, _this._updateConfig = function (_ref2) {
      var key = _ref2.key,
          value = _ref2.value;

      var currentValue = _this.state.config[key].value;
      if (value !== currentValue) {
        _this.setState({
          config: (0, _extends4.default)({}, _this.state.config, (0, _defineProperty3.default)({}, key, (0, _extends4.default)({}, _this.state.config[key], {
            value: value
          })))
        });
      }
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ColorRangeSelect, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var config = this.state.config;

      return _react2.default.createElement(
        ColorRangeSelector,
        { className: 'color-range-selector' },
        _react2.default.createElement(
          StyledColorConfig,
          null,
          Object.keys(config).map(function (key) {
            return _react2.default.createElement(PaletteConfig, {
              key: key,
              label: key,
              config: config[key],
              onChange: function onChange(value) {
                return _this2._updateConfig({ key: key, value: value });
              }
            });
          })
        ),
        _react2.default.createElement(ColorPaletteGroup, {
          config: config,
          colorRanges: this.props.colorRanges,
          onSelect: this.props.onSelectColorRange,
          selected: this.props.selectedColorRange
        })
      );
    }
  }]);
  return ColorRangeSelect;
}(_react.Component), _class.propTypes = {
  colorRanges: _propTypes2.default.arrayOf(_propTypes2.default.any),
  selectedColorRange: _propTypes2.default.object,
  onSelectColorRange: _propTypes2.default.func.isRequired
}, _class.defaultProps = {
  colorRanges: _colorRanges.COLOR_RANGES,
  onSelectColorRange: function onSelectColorRange() {}
}, _temp2);
exports.default = ColorRangeSelect;


var StyledPaletteConfig = _styledComponents2.default.div(_templateObject3, function (props) {
  return props.theme.secondaryInput;
});

var PaletteConfig = function PaletteConfig(_ref3) {
  var category = _ref3.category,
      label = _ref3.label,
      _ref3$config = _ref3.config,
      type = _ref3$config.type,
      value = _ref3$config.value,
      options = _ref3$config.options,
      _onChange = _ref3.onChange;
  return _react2.default.createElement(
    StyledPaletteConfig,
    {
      className: 'color-palette__config',
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    },
    _react2.default.createElement(
      'div',
      { className: 'color-palette__config__label' },
      _react2.default.createElement(
        _styledComponents3.PanelLabel,
        null,
        label
      )
    ),
    type === 'select' && _react2.default.createElement(
      'div',
      { className: 'color-palette__config__select' },
      _react2.default.createElement(_itemSelector2.default, {
        selectedItems: value,
        options: options,
        multiSelect: false,
        searchable: false,
        onChange: _onChange
      })
    ),
    type === 'slider' && _react2.default.createElement(
      'div',
      { className: 'color-palette__config__slider' },
      _react2.default.createElement(
        'div',
        { className: 'color-palette__config__slider__slider' },
        _react2.default.createElement(_rangeSlider2.default, {
          range: options,
          value0: options[0],
          value1: value,
          step: 1,
          isRanged: false,
          showInput: false,
          onChange: function onChange(val) {
            return _onChange(val[1]);
          }
        })
      ),
      _react2.default.createElement(
        'div',
        { className: 'color-palette__config__slider__number' },
        value
      )
    ),
    type === 'switch' && _react2.default.createElement(_switch2.default, {
      checked: value,
      id: category + '-' + label + '-toggle',
      onChange: function onChange() {
        return _onChange(!value);
      },
      secondary: true
    })
  );
};

var StyledColorRange = _styledComponents2.default.div(_templateObject4, function (props) {
  return props.theme.panelBackgroundHover;
});

var ColorPaletteGroup = function ColorPaletteGroup(_ref4) {
  var _ref4$config = _ref4.config,
      config = _ref4$config === undefined ? {} : _ref4$config,
      onSelect = _ref4.onSelect,
      selected = _ref4.selected,
      colorRanges = _ref4.colorRanges;
  var steps = config.steps,
      reversed = config.reversed,
      type = config.type;


  var filtered = colorRanges.filter(function (colorRange) {
    var isType = !type || type.value === 'all' || type.value === colorRange.type;
    var isStep = !steps || Number(steps.value) === colorRange.colors.length;

    return isType && isStep;
  });

  var isReversed = Boolean(reversed && reversed.value);

  return _react2.default.createElement(
    'div',
    { className: 'color-palette__group' },
    filtered.map(function (colorRange) {
      return _react2.default.createElement(
        StyledColorRange,
        {
          className: 'color-ranges',
          key: colorRange.name,
          onClick: function onClick(e) {
            return onSelect((0, _extends4.default)({}, colorRange, {
              reversed: isReversed,
              colors: isReversed ? colorRange.colors.slice().reverse() : colorRange.colors
            }), e);
          }
        },
        _react2.default.createElement(_colorPalette2.default, {
          colors: colorRange.colors,
          isReversed: isReversed,
          isSelected: colorRange.name === selected.name && isReversed === Boolean(selected.reversed)
        })
      );
    })
  );
};
