// kepler.gl/dist/deckgl-layers/h3-hexagon-cell-layer/h3-hexagon-cell-layer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _get2 = require('babel-runtime/helpers/get');

var _get3 = _interopRequireDefault(_get2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _deck = require('deck.gl');

var _luma = require('luma.gl');

var _h3Utils = require('../../layers/h3-hexagon-layer/h3-utils');

var _shaderUtils = require('../layer-utils/shader-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

function addInstanceCoverage(vs) {
  var addDecl = (0, _shaderUtils.editShader)(vs, 'hexagon cell vs add instance', 'attribute vec3 instancePickingColors;', 'attribute vec3 instancePickingColors;\n     attribute float instanceCoverage;');

  return (0, _shaderUtils.editShader)(addDecl, 'hexagon cell vs add instance', 'float dotRadius = project_scale(radius) * mix(coverage, 0.0, noRender);', 'float dotRadius = project_scale(radius) * mix(coverage * instanceCoverage, 0.0, noRender);');
}

// TODO: export all dekc.gl layers from kepler.gl

var H3HexagonCellLayer = function (_HexagonCellLayer) {
  (0, _inherits3.default)(H3HexagonCellLayer, _HexagonCellLayer);

  function H3HexagonCellLayer() {
    (0, _classCallCheck3.default)(this, H3HexagonCellLayer);
    return (0, _possibleConstructorReturn3.default)(this, (H3HexagonCellLayer.__proto__ || Object.getPrototypeOf(H3HexagonCellLayer)).apply(this, arguments));
  }

  (0, _createClass3.default)(H3HexagonCellLayer, [{
    key: 'getShaders',
    value: function getShaders() {
      var shaders = (0, _get3.default)(H3HexagonCellLayer.prototype.__proto__ || Object.getPrototypeOf(H3HexagonCellLayer.prototype), 'getShaders', this).call(this);

      return (0, _extends3.default)({}, shaders, {
        vs: addInstanceCoverage(shaders.vs)
      });
    }
  }, {
    key: 'getCylinderGeometry',
    value: function getCylinderGeometry(radius) {
      var distortion = this.getDistortion();

      var cylinderGeometry = new _luma.CylinderGeometry({
        radius: radius,
        topRadius: radius,
        bottomRadius: radius,
        topCap: true,
        bottomCap: true,
        height: 1,
        verticalAxis: 'z',
        nradial: 6,
        nvertical: 1
      });

      if (distortion) {
        var pos = cylinderGeometry.attributes.positions.value;
        var adjusted = (0, _h3Utils.distortCylinderPositions)(pos, distortion);
        cylinderGeometry.attributes.positions.value = adjusted;
      }

      return cylinderGeometry;
    }
  }, {
    key: 'getDistortion',
    value: function getDistortion() {
      var _this2 = this;

      var _props = this.props,
          hexagonVertices = _props.hexagonVertices,
          hexagonCenter = _props.hexagonCenter;


      if (Array.isArray(hexagonVertices) && hexagonVertices.length >= 6 && Array.isArray(hexagonCenter)) {
        var screenVertices = hexagonVertices.map(function (d) {
          return _this2.projectFlat(d);
        });
        var screenCentroid = this.projectFlat(hexagonCenter);
        return (0, _h3Utils.getH3VerticeTransform)(screenVertices, screenCentroid);
      }

      return null;
    }
  }, {
    key: 'updateRadiusAngle',
    value: function updateRadiusAngle() {
      var _props2 = this.props,
          angle = _props2.angle,
          radius = _props2.radius;
      var hexagonVertices = this.props.hexagonVertices;


      if (Array.isArray(hexagonVertices) && hexagonVertices.length >= 6) {
        var viewport = this.context.viewport;
        // calculate angle and vertices from hexagonVertices if provided

        var vertices = this.props.hexagonVertices;

        var vertex0 = vertices[0];
        var vertex3 = vertices[3];

        // project to space coordinates

        var _viewport$getDistance = viewport.getDistanceScales(),
            pixelsPerMeter = _viewport$getDistance.pixelsPerMeter;

        var spaceCoord0 = this.projectFlat(vertex0);
        var spaceCoord3 = this.projectFlat(vertex3);

        angle = (0, _h3Utils.getAngle)(spaceCoord0, spaceCoord3);
        radius = (0, _h3Utils.getRadius)(spaceCoord0, spaceCoord3) / pixelsPerMeter[0];
      }

      this.setState({ angle: angle, radius: radius });
    }
  }, {
    key: 'draw',
    value: function draw(opts) {
      var uniforms = opts.uniforms;


      (0, _get3.default)(H3HexagonCellLayer.prototype.__proto__ || Object.getPrototypeOf(H3HexagonCellLayer.prototype), 'draw', this).call(this, (0, _extends3.default)({}, opts, {
        uniforms: (0, _extends3.default)({}, uniforms, {
          picking_uHighlightScale: this.props.extruded ? 1.4 : 0.0
        })
      }));
    }
  }, {
    key: 'initializeState',
    value: function initializeState() {
      (0, _get3.default)(H3HexagonCellLayer.prototype.__proto__ || Object.getPrototypeOf(H3HexagonCellLayer.prototype), 'initializeState', this).call(this);

      this.getAttributeManager().addInstanced({
        instanceCoverage: { size: 1, accessor: 'getCoverage' }
      });
    }
  }]);
  return H3HexagonCellLayer;
}(_deck.HexagonCellLayer);

exports.default = H3HexagonCellLayer;


H3HexagonCellLayer.layerName = 'H3HexagonCellLayer';
